### PR TITLE
[SofaCore] Add PCH support in CMakeLists.txt

### PR DESCRIPTION
**Pull request by damienmarchal**
Using clang -ftimetrace show that BaseObject/Data/Base are taking a lot of compilation time
because of excessive inclusion and parsing time.
PCH can help to solve this issue and there is support for that in CMake.
Let's se result.

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Comment by damienmarchal** [comment]
Sadly the CI is using cmake 3.12.
PCH support starts at 3.16 
https://cmake.org/cmake/help/latest/command/target_precompile_headers.html

File: SofaKernel/modules/SofaCore/CMakeLists.txt [hunk 1]
```diff
@@ -252,6 +252,11 @@ set(SOURCE_FILES
 add_library(${PROJECT_NAME} SHARED ${HEADER_FILES} ${SOURCE_FILES})
 target_link_libraries(${PROJECT_NAME} PUBLIC SofaHelper SofaDefaultType)
 
+if((NOT DISABLE_PRECOMPILE_HEADERS) AND (${CMAKE_VERSION} VERSION_GREATER "3.16.0"))
```
Comment (guparan) [inline SofaKernel/modules/SofaCore/CMakeLists.txt]: `if(SOFA_BUILD_WITH_PCH_ENABLED)` would be better

File: CMakeLists.txt [hunk 2]
```diff
@@ -27,6 +27,16 @@ set_property(GLOBAL PROPERTY USE_FOLDERS ON)
...
 
+# Option to accelerate the building
+# https://cmake.org/cmake/help/v3.19/command/target_precompile_headers.html
+option(SOFA_BUILD_WITH_PCH_ENABLED "Compile Sofa using precompiled header (to accelerate the build process)" OFF)
+if(SOFA_BUILD_WITH_PCH_ENABLED)
+    message("-- Precompiled headers: Yes (SOFA_BUILD_WITH_PCH_ENABLED variable is ON).")
+else()
+    message("-- Precompiled headers: No (SOFA_BUILD_WITH_PCH_ENABLED variable is OFF).")
+    set(DISABLE_PRECOMPILE_HEADERS ON)   # https://cmake.org/cmake/help/v3.19/prop_tgt/DISABLE_PRECOMPILE_HEADERS.html
+endif()
```
Comment (guparan) [inline CMakeLists.txt]:
```cmake
if(CMAKE_VERSION VERSION_GREATER "3.16.0")
    option(SOFA_BUILD_WITH_PCH_ENABLED "Compile Sofa using precompiled header (to accelerate the build process)" OFF)
    if(SOFA_BUILD_WITH_PCH_ENABLED)
        message("-- Precompiled headers: Yes (SOFA_BUILD_WITH_PCH_ENABLED is ON).")
    else()
        message("-- Precompiled headers: No (SOFA_BUILD_WITH_PCH_ENABLED is OFF).")
        set(DISABLE_PRECOMPILE_HEADERS ON)   # https://cmake.org/cmake/help/v3.19/prop_tgt/DISABLE_PRECOMPILE_HEADERS.html
    endif()
endif()
```

**Review by guparan — CHANGES_REQUESTED** [review]

**Comment by damienmarchal** [comment]
I Updated the change to take into account's guillaume comments

File: CMakeLists.txt (hunk 2 above)
Comment (damienmarchal) [reply CMakeLists.txt]: Done

**Pull request merged**